### ISSUE TITLE
[a11y] VoiceOver does not announce state of ButtonSegment

### ISSUE THREAD

**Issue by scopendo** [flutter/flutter#146987]
### Steps to reproduce

Create a widget that shows a `SegmentedButton`, without any explicit use of `Semantics`.  The example below is for a light/dark mode toggle.

```dart
class _ThemeSelector extends StatelessWidget {
  const _ThemeSelector();

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    // Get the current theme mode...
    final themeMode = ...;

    return SegmentedButton(
      showSelectedIcon: false,
      segments: [
        ButtonSegment(
          icon: const Icon(Icons.light_mode),
          label: Text('Light mode'),
          value: ThemeMode.light,
        ),
        ButtonSegment(
          icon: const Icon(Icons.dark_mode),
          label: Text('Dark mode'.hardcoded),
          value: ThemeMode.dark,
        ),
      ],
      selected: {themeMode},
      onSelectionChanged: (selected) {
        // Set the current theme mode...
      },
    );
  }
}
```

### Expected results

I would expect that the button segment announces that it is selected, or perhaps toggled on, without needing to wrap `ButtonSegment.label` in an explicit `Semantics` widget.

For example, given the example, it should announce "selected, light mode, button".

It is also probable that the role of a button segment is a toggle, rather than a button.

### Actual results

In the example above, VoiceOver announces the button segment's label followed by "button" but without indicating when the button segment is selected, for example, "light mode, button".

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Segmented Button Demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  var _themeMode = ThemeMode.light;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: Text(widget.title),
      ),
      body: Center(
        child: SegmentedButton(
          showSelectedIcon: false,
          segments: const [
            ButtonSegment(
              icon: Icon(Icons.light_mode),
              label: Text('Light mode'),
              value: ThemeMode.light,
            ),
            ButtonSegment(
              icon: Icon(Icons.dark_mode),
              label: Text('Dark mode'),
              value: ThemeMode.dark,
            ),
          ],
          selected: {_themeMode},
          onSelectionChanged: (selected) {
            setState(() {
              _themeMode = selected.first;
            });
          },
        ),
      ),
    );
  }
}
```

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details open><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
Doctor summary (to see all details, run flutter
doctor -v):
[✓] Flutter (Channel stable, 3.19.5, on macOS 14.2.1 23C71 darwin-arm64, locale en-GB)
[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
[✓] Xcode - develop for iOS and macOS (Xcode 15.3)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2023.2)
[✓] IntelliJ IDEA Ultimate Edition (version 2023.3.6)
[✓] Connected device (3 available)
[✓] Network resources

• No issues found!
```

</details>

**Comment by huycozy** [comment]
Thanks for the report. Reproduced the issue on iOS (with VoiceOver). 

On Android (with TalkBack), it seems to work as expected when the buttons are announced "checked" state, for e.g:
"checked, light mode, radio button" and "not checked, dark mode, radio button"


<details>
<summary>Demo</summary>

https://github.com/flutter/flutter/assets/104349824/7e4460a9-4eef-4839-9ce7-85b14ae2bb74

</details>

<details>
<summary>flutter doctor -v (stable and master)</summary>

```bash
[✓] Flutter (Channel stable, 3.19.6, on macOS 14.1 23B74 darwin-x64, locale en-VN)
    • Flutter version 3.19.6 on channel stable at /Users/huynq/Documents/GitHub/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 54e66469a9 (31 hours ago), 2024-04-17 13:08:03 -0700
    • Engine revision c4cd48e186
    • Dart version 3.3.4
    • DevTools version 2.31.1

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.9+0-17.0.9b1087.7-11185874)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.3)
    • Xcode at /Applications/Xcode15.3.app/Contents/Developer
    • Build 15E204a
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2023.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • android-studio-dir = /Applications/Android Studio.app/
    • Java version OpenJDK Runtime Environment (build 17.0.9+0-17.0.9b1087.7-11185874)

[✓] VS Code (version 1.88.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.86.0

[✓] Connected device (3 available)
    • RMX2001 (mobile) • EUYTFEUSQSRGDA6D • android-arm64  • Android 11 (API 30)
    • macOS (desktop)  • macos            • darwin-x64     • macOS 14.1 23B74 darwin-x64
    • Chrome (web)     • chrome           • web-javascript • Google Chrome 123.0.6312.124

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

```bash
[!] Flutter (Channel master, 3.22.0-14.0.pre.31, on macOS 14.1 23B74 darwin-x64, locale en-VN)
    • Flutter version 3.22.0-14.0.pre.31 on channel master at /Users/huynq/Documents/GitHub/flutter_master
    ! Warning: `flutter` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/flutter, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_master. Consider adding /Users/huynq/Documents/GitHub/flutter_master/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/dart, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_master. Consider adding /Users/huynq/Documents/GitHub/flutter_master/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 20034ce0a2 (39 minutes ago), 2024-04-18 22:33:19 -0400
    • Engine revision ff471881e9
    • Dart version 3.5.0 (build 3.5.0-75.0.dev)
    • DevTools version 2.35.0-dev.8
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.9+0-17.0.9b1087.7-11185874)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.3)
    • Xcode at /Applications/Xcode15.3.app/Contents/Developer
    • Build 15E204a
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2023.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • android-studio-dir = /Applications/Android Studio.app/
    • Java version OpenJDK Runtime Environment (build 17.0.9+0-17.0.9b1087.7-11185874)

[✓] VS Code (version 1.88.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.86.0

[✓] Connected device (3 available)
    • RMX2001 (mobile) • EUYTFEUSQSRGDA6D • android-arm64  • Android 11 (API 30)
    • macOS (desktop)  • macos            • darwin-x64     • macOS 14.1 23B74 darwin-x64
    • Chrome (web)     • chrome           • web-javascript • Google Chrome 123.0.6312.124

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```
</details>

**Comment by scopendo** [comment]
@huycozy – I don't think the issue should specifically be about announcing "selected", which is how I read the updated title, as it really should not be announced as a button, as you note for Android (though for me it announces "ticked, light mode, radio button" / "not ticked, dark mode, radio button".

**Comment by huycozy** [comment]
Hmm, I heard it as "checked" on Android. It will be more clearly if we look at the code. But anyway, I will re-title this. Thanks!

<details>
<summary>Demo Android</summary>

https://github.com/flutter/flutter/assets/104349824/9634cc47-746f-4c9e-b6f8-55e113a8bde5
</details>

**Comment by chunhtai** [comment]
cc @QuncCccccc  looks like the segmented buttons need to be wrapped with Semantics(selected:true) for the button that is selected

**Comment by QuncCccccc** [comment]
> cc @QuncCccccc looks like the segmented buttons need to be wrapped with Semantics(selected:true) for the button that is selected

Adding `selected` seems not working. Seems `RadioButton` also set `selected` in `Semantics` but VoiceOver doesn't announce selected or not as well.

**Comment by hannah-hyj** [comment]
On web this is read out as "radio button, selected" if it's selected. 

So it looks like this is only an iOS issue.  
lowering the priority since it's iOS only

**Comment by kevmoo** [comment]
@hannah-hyj – what's the screen reader being used for your verification? Is it showing up okay w/ VoiceOver (mac)?

**Comment by hannah-hyj** [comment]
@kevmoo  I'm using voiceover on mac

**Comment by hannah-hyj** [comment]
@kevmoo verified with Yegor this is working on windows with JAWS and NVDA.

**Comment by huycozy** [comment]
I tried to fix this issue, but I couldn't get the existing tests to pass. Therefore, I am unassigning myself (I apologize for any confusion this may cause). I believe the solution should be here:

From:
https://github.com/flutter/flutter/blob/7130981fe1ca12481f0bd5c4bdf6dbcee84afe43/packages/flutter/lib/src/material/segmented_button.dart#L632-L638

To:

```dart
return MergeSemantics(
        child: Semantics(
          selected: segmentSelected,
          inMutuallyExclusiveGroup: widget.multiSelectionEnabled ? null : true,
          child: buttonWithTooltip,
        ),
      );
```

It's replacing `checked` to `selected`, then the state can be announced as expected. It's `replacing` action but not `adding` because `adding` will cause the button to be read with both `checked` and `selected` states with TalkBack reader (tested this).

**Comment by QuncCccccc** [comment]
> I tried to fix this issue, but I couldn't get the existing tests to pass. 

Do you still remember which tests will fail?

**Comment by huycozy** [comment]
> Do you still remember which tests will fail?

@QuncCccccc Yes, they are here, but when I try to reproduce the failure again, I fortunately managed to resolve them. Perhaps I was a little confused earlier. I will try again. 😅

https://github.com/flutter/flutter/blob/9d4db1d5dd5682dcb459b3a05ae4e16d102e1e80/packages/flutter/test/material/segmented_button_test.dart#L468

https://github.com/flutter/flutter/blob/9d4db1d5dd5682dcb459b3a05ae4e16d102e1e80/packages/flutter/test/material/segmented_button_test.dart#L542
